get_started_with_membrane/03_elements.md

```diff
@@ -0,0 +1,276 @@
+    The factor by which the volume will be changed.
+
+    A gain smaller than 1 reduces the volume and gain
+    greater than 1 increases it.
+    """
+  ]
+
+  @impl true
+  def handle_init(_ctx, options) do
+    {[], %{gain: options.gain}}
+  end
+  
+  @impl true
+  def handle_process(:input, buffer, ctx, state) do
+    stream_format = ctx.pads.input.stream_format
+    sample_size = RawAudio.sample_size(stream_format)
+    payload =
+      for <<sample::binary-size(sample_size) <- buffer.payload>>, into: <<>> do
+        value = RawAudio.sample_to_value(sample, stream_format)
+        scaled_value = round(value * state.gain)
+        RawAudio.value_to_sample(scaled_value, stream_format)
+      end
+
+    buffer = %Membrane.Buffer{buffer | payload: payload}
+    {[buffer: {:output, buffer}], state}
+  end
+end
+```
+
+As the `moduledoc` says, the element can be used to adjust the audio volume. As we create a filter, we start with `use Membrane.Filter` clause. Then we define pads, one input and one output:
+
+```elixir
+alias Membrane.RawAudio
+
+def_input_pad :input, accepted_format: RawAudio, flow_control: :auto
+def_output_pad :output, accepted_format: RawAudio, flow_control: :auto
+```
+
+The element is going to receive raw audio and send the raw audio too. The raw audio (sometimes referred to as PCM - Pulse Code Modulation) is a simple digital representation of an audio wave, that we can operate on - for example, change the volume. The `Membrane.RawAudio` format is defined in the `membrane_raw_audio_format` package.
+
+Since the element only transforms the stream as it flows, we can safely set `flow_control` to `auto` on both pads.
+
+After defining the pads, we can define options. In this case, it's a single option - `gain` by which the volume will be changed.
+
+```elixir
+def_options gain: [
+  spec: number(),
+  description: """
+  The factor by which the volume will be changed.
+
+  A gain smaller than 1 reduces the volume and gain
+  greater than 1 increases it.
+  """
+]
+```
+
+It's important to provide the type spec and description for each option so that everyone knows how to use it.
+
+Next, we implement the first callback - `handle_init`:
+
+```elixir
+@impl true
+def handle_init(_ctx, options) do
+  {[], %{gain: options.gain}}
+end
+```
+
+The callback does not return any actions (thus the empty list), but it saves the gain passed through options in the state.
+
+Then goes the main part of the element - the `handle_process` callback:
+
+```elixir
+@impl true
+def handle_process(:input, buffer, ctx, state) do
+```
+
+The callback is called whenever a buffer arrives on a pad, and receives four arguments:
+- the pad where the buffer arrived,
+- the [Membrane.Buffer](https://hexdocs.pm/membrane_core/Membrane.Buffer.html) structure carrying the stream data,
+- [Membrane.Element.CallbackContext](https://hexdocs.pm/membrane_core/Membrane.Element.CallbackContext.html), providing some useful information about the element,
```

Comment:
```suggestion
- [Membrane.Element.CallbackContext.t()](https://hexdocs.pm/membrane_core/Membrane.Element.CallbackContext.html#t:t/0), providing some useful information about the element,
```

Reply:
is there any difference?

Reply:
typespec vs module

Reply:
Especially this typespec describe a map, not a structure